pkg/trace/api/evp_intake_proxy.go

```diff
@@ -0,0 +1,233 @@
+func evpProxyErrorHandler(message string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		msg := fmt.Sprintf("EVPProxy is disabled: %v", message)
+		http.Error(w, msg, http.StatusMethodNotAllowed)
+	})
+}
+
+// evpProxyForwarder creates an http.ReverseProxy which can forward payloads to
+// one or more endpoints, based on the request received and the Agent configuration.
+// Headers are not proxied, instead we add our own known set of headers.
+// See also evpProxyTransport below.
+func evpProxyForwarder(conf *config.AgentConfig, endpoints []config.Endpoint, transport http.RoundTripper, logger *stdlog.Logger) http.Handler {
+	director := func(req *http.Request) {
+
+		containerID := req.Header.Get("Datadog-Container-ID")
+		contentType := req.Header.Get("Content-Type")
+		userAgent := req.Header.Get("User-Agent")
+
+		// Clear all received headers
+		req.Header = http.Header{}
+
+		// Standard headers
+		req.Header.Set("Via", fmt.Sprintf("trace-agent %s", info.Version))
+		if contentType != "" {
+			req.Header.Set("Content-Type", contentType)
+		}
+		req.Header.Set("User-Agent", userAgent) // Set even if an empty string so Go doesn't set its default
+		req.Header["X-Forwarded-For"] = nil     // Prevent setting X-Forwarded-For
+
+		// Datadog headers
+		if ctags := getContainerTags(conf.ContainerTags, containerID); ctags != "" {
+			req.Header.Set("X-Datadog-Container-Tags", ctags)
+		}
+		req.Header.Set("X-Datadog-Hostname", conf.Hostname)
+		req.Header.Set("X-Datadog-AgentDefaultEnv", conf.DefaultEnv)
+
+		// URL, Host and the API key header are set in the transport for each outbound request
+	}
+
+	return &httputil.ReverseProxy{
+		Director:  director,
+		ErrorLog:  logger,
+		Transport: &evpProxyTransport{transport, endpoints},
+	}
+}
+
+// evpProxyTransport sends HTTPS requests to multiple targets using an
+// underlying http.RoundTripper. API keys are set separately for each target.
+// When multiple endpoints are in use the response from the first endpoint
+// is proxied back to the client, while for all aditional endpoints the
+// response is discarded.
+type evpProxyTransport struct {
+	transport http.RoundTripper
+	endpoints []config.Endpoint
+}
+
+func (t *evpProxyTransport) RoundTrip(req *http.Request) (rresp *http.Response, rerr error) {
+
+	// Metrics with stats for debugging
+	beginTime := time.Now()
+	metricTags := []string{}
+	defer func() {
+		metrics.Count("datadog.trace_agent.evp_proxy.request", 1, metricTags, 1)
+		metrics.Count("datadog.trace_agent.evp_proxy.request_bytes", req.ContentLength, metricTags, 1)
+		metrics.Timing("datadog.trace_agent.evp_proxy.request_duration_ms", time.Since(beginTime), metricTags, 1)
+		if rerr != nil {
+			metrics.Count("datadog.trace_agent.evp_proxy.request_error", 1, metricTags, 1)
+		}
+	}()
+
+	// Parse request path: The first component is the target subdomain, the rest is the target path.
+	inputPath := req.URL.Path
```

Comment:
req and req.URL can be nil and should be tested as execution preconditions to prevent a crash if the URL is invalid via query header manipulation.

Reply:
I don't think `req` can be null: If it were, it would crash way before our code, in the reverseproxy code: https://github.com/golang/go/blob/dev.boringcrypto.go1.17/src/net/http/httputil/reverseproxy.go#L234

About `req.URL` I'm less sure, but I see no null checks on Go's own handlers like StripPrefix here: https://github.com/golang/go/blob/770e0e584a98dfd5e8d0d00558085c339fda0ed7/src/net/http/server.go#L2138

Reply:
I don't think about the fact that a pointer can't be nil or not due to parent caller nil handling. By design, a pointer can be nil and must be tested when it's coming from a different scope for a defensive posture. In go, there is no pointer reference borrowing management like in Rust. Concerning the `req.URL` to be nil, it happens when you use a custom router (like gin/echo) where the URL parsing is done externally to build the `req` instance. I had bad situations to handle due to that :-).

Reply:
Added null checks.

Reply:
I disagree. By design `req` should not be `nil`. If it is ever `nil`, then that's a big programming error or an intentional malice which deserves a `panic`. We don't do this anywhere so let's not start here please.

Reply:
I also think it cannot be null (it would blow up in Go's code before reaching our code) but I didn't fully understand the case @Zenithar was talking about so I just added the checks. Happy to revise, though. And to be fair we don't do this in any other endpoint.